pkg/metrics/collector/kernel_metrics.go

```diff
@@ -0,0 +1,158 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	tcpTimeoutDesc = prometheus.NewDesc(
+		"tcp_timeout_rehash_count",
+		"Tcp timeout rehash, collected from nstat",
+		nil, nil,
+	)
+	tcpDuplicateDesc = prometheus.NewDesc(
+		"tcp_duplicate_rehash_count",
+		"Tcp duplicate rehash, collected from nstat",
+		nil, nil,
+	)
+	segmentsReceivedDesc = prometheus.NewDesc(
+		"tcp_segments_received_count",
+		"TCP segments received on the node, collected from netstat -s",
+		nil, nil,
+	)
+	segmentsSentDesc = prometheus.NewDesc(
+		"tcp_segments_sent_count",
+		"TCP segments sent from node, collected from netstat -s",
+		nil, nil,
+	)
+	segmentsRetransmittedDesc = prometheus.NewDesc(
+		"tcp_segments_retransmitted_count",
+		"TCP segments retransmitted on the node, collected from netstat -s",
+		nil, nil,
+	)
+)
+
+type kernelStatCollector struct {
+}
+
+func init() {
+	registerCollector("kernel_metrics", NewKernelStatCollector)
+}
+
+func NewKernelStatCollector() (Collector, error) {
+	return &kernelStatCollector{}, nil
+}
+
+// parseNetstatVals parses the contents of /proc/net/netstat
+// looking for lines such as:
+//
+// TcpExt: SyncookiesSent RcvPruned OfoPruned OutOfWindowIcmps LockDroppedIcmps TcpTimeoutRehash TcpDuplicateDataRehash TCPDSACKRecvSegs TCPDSACKIgnoredDubious
+// TcpExt: 0 0 0 0 0 440 99 21800 3
+// and populates a map with all the parsed values where the keys are
+// the labels in the first line, separated by spaces and second line contains the values,
+// indexed accordingly
+func parseNetstatVals(output string) map[string]uint64 {
+	stats := make(map[string]uint64)
+	splitFile := strings.Split(output, "\n")
+	for i, stat := range splitFile {
+
+		if !strings.Contains(stat, "TcpExt:") {
+			continue
+		}
+		keyList := strings.Fields(stat)
+		valList := strings.Fields(splitFile[i+1])
+		for i, key := range keyList[1:] {
+			val, err := strconv.Atoi(valList[i+1])
+			if err != nil {
+				continue
+			}
+			stats[key] = uint64(val)
+		}
+		return stats
+	}
+	return stats
+}
+
+// parseSnmpVals parses the contents of /proc/net/snmp
+// looking for lines such as:
+//
+// Tcp: RtoAlgorithm RtoMin RtoMax MaxConn ActiveOpens PassiveOpens AttemptFails EstabResets CurrEstab InSegs OutSegs RetransSegs InErrs OutRsts InCsumErrors
+// Tcp: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
+//
+// and populates a map with all the parsed values where the key are the labels in the first line separated by spaces
+// and the values are the numbers in the second line, casted to uint64
+
+func parseSnmpVals(output string) map[string]uint64 {
```

Comment:
This looks like exactly the same function as above. Maybe add a helper that takes in the two lines and returns the map.

Reply:
Done, have a function that takes the label which is used in the update function